Sandbox extension to container temporary directory is not created for non persistent Website data stores

pvollan opened the pull request:
#### ada763c496ca7c696e88ea4fa756db416d7e07ce
<pre>
Sandbox extension to container temporary directory is not created for non persistent Website data stores
<a href="https://bugs.webkit.org/show_bug.cgi?id=242946">https://bugs.webkit.org/show_bug.cgi?id=242946</a>
&lt;rdar://95817773&gt;

Reviewed by Chris Dumez.

For persistent Website data stores we create a sandbox extension to the container temporary directory for the WebContent process.
This is also needed for non persistent Website data stores, since other frameworks may depend on access to this directory. For
example, image decoding depends on having write access to this location. This patch removes the check for persistent data store
in WebsiteDataStore::resolvedContainerTemporaryDirectory. This method is only used when creating the sandbox extension, so the
only behavior change is that we will now get a sandbox extension also for non persistent data stores.

* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::resolvedContainerTemporaryDirectory):
(WebKit::WebsiteDataStore::defaultResolvedContainerTemporaryDirectory):
(WebKit::WebsiteDataStore::defaultContainerTemporaryDirectory): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: <a href="https://commits.webkit.org/252676@main">https://commits.webkit.org/252676@main</a>
</pre>

cdumez reviewed (approved):
OK since I think this restores earlier behavior. However, I think that ideally, a WebContent process from an ephemeral data store wouldn't have to write to disk *ever*, even to a temporary folder.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -684,19 +684,20 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
+    static NeverDestroyed<String> resolvedTemporaryDirectory;
+    static std::once_flag once;
+    std::call_once(once, [] {
+        resolvedTemporaryDirectory.get() = resolveAndCreateReadWriteDirectoryForSandboxExtension(String(NSTemporaryDirectory()));
```
Do we really need the explicit `String()`?

pvollan replied:
Yes, we currently need that.

Thanks for reviewing!

szewai replied:
Nice, this should reduce some calls to resolve for different WebsiteDataStores

pvollan replied:
Yes, I think this should reduce the number of resolve calls.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -684,19 +684,20 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
+    std::call_once(once, [] {
+        resolvedTemporaryDirectory.get() = resolveAndCreateReadWriteDirectoryForSandboxExtension(String(NSTemporaryDirectory()));
+    });
+    return resolvedTemporaryDirectory.get();
```
You probably don't need `.get()`

pvollan committed:
Sandbox extension to container temporary directory is not created for non persistent Website data stores
https://bugs.webkit.org/show_bug.cgi?id=242946
<rdar://95817773>

Reviewed by Chris Dumez.

For persistent Website data stores we create a sandbox extension to the container temporary directory for the WebContent process.
This is also needed for non persistent Website data stores, since other frameworks may depend on access to this directory. For
example, image decoding depends on having write access to this location. This patch removes the check for persistent data store
in WebsiteDataStore::resolvedContainerTemporaryDirectory. This method is only used when creating the sandbox extension, so the
only behavior change is that we will now get a sandbox extension also for non persistent data stores.

* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::resolvedContainerTemporaryDirectory):
(WebKit::WebsiteDataStore::defaultResolvedContainerTemporaryDirectory):
(WebKit::WebsiteDataStore::defaultContainerTemporaryDirectory): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: https://commits.webkit.org/252676@main
